feat(log): add deriveLogger() to simplify derived loggers

tripodsan opened the pull request:
- adds new option to logger and interface base constructor to specify constant log fields (`defaultFields`)
- adds a new function: `deriveLogger<T>(T: logger, object: opts): T`

koraa commented:
Could you please elaborate on why this feature specifically is needed? Seems like a fairly specific feature that could be accomplished with the `filter` property?

I think – if we decide we would like to go ahead with such a feature – we would also think about how to make the api consistent. E.g.  why are there now two ways of adding fields? Why is this method only implemented for SimpleInterface and not the other loggers & interfaces? Why can the fields property be mutated using `.child()` but not filter, logger, level…

This might be better suited as a helper function, as that would work on any logger/interface and requirements in terms of API consistency would be much lower:

```js
const withFields = (logger, fields) => {
  // We should probably consider just implementing the Deepclone/Shallowclone traits…
  const nu = Object.assign(Object.create({}, logger.constructor), logger); // shallowclone
  nu.filter = (msg) => nu.filter({...fields, ...msg}); // function composition (monadically extend filter)
  return nu;
};
```

tripodsan commented:
> Could you please elaborate on why this feature specifically is needed? 

because it's quite useful to create derived loggers that you can pass to sub-components that include additional fields. eg:

```
const requestLogger = log.child({ request });
processRequest(requestLogger, request);
```

> Seems like a fairly specific feature that could be accomplished with the `filter` property?

I think the filter should filter whatever the log system provides as fields.

> I think – if we decide we would like to go ahead with such a feature – we would also think about how to make the api consistent. E.g. why are there now two ways of adding fields?

well, one way is to add fields for 1 message. but the child logger offers the ability the create a derived logger with constant fields.

> Why is this method only implemented for SimpleInterface and not the other loggers & interfaces? 

good point... I'll move it to the base interface

> Why can the fields property be mutated using `.child()` but not filter, logger, level…

good point...

> This might be better suited as an ad-hoc helper function in your repo, as that would work on any logger/interface and requirements in terms of API consistency would be much lower:
> 
> ```js
> const withFields = (logger, fields) => {
>   // We should probably consider just implementing the Deepclone/Shallowclone traits…
>   const nu = Object.assign(Object.create({}, logger.constructor), logger); // shallowclone
>   nu.filter = (msg) => nu.filter({...fields, ...msg}); // function composition (monadically extend filter)
>   return nu;
> };
> ```

if we allow to create derived loggers for all "fields", then the filter is not enough.

tripodsan commented:
@koraa I moved the `child` method and allow for generically create sub loggers. But I also added the ability to specify _base fields_.

koraa commented:
Thanks for the improvement. This is definitely better!

the issue of why this method is only available for SimpleLogger is not addressed; in addition to that the feature became even more specific; the deep merging aspect is one that might be surprising and cause issues. It is certainly useful in some cases, but should definitely be opt in. And again, adding fields using the filter property is very easy; I do not see any reason why it would be necessary to make a fields property a first class concept.

`child()` should really be generically available and it can be implemented as a function. I generally won't accept a PR that just introduces a generic concept for just one of the classes; I also won't be extending the existing interfaces/traits when a function will do. The traits should stay as simple as possible so it stays as easy to implement.

However, you have convinced me that creating derived loggers is a useful feature that should be added to helix log; it's just a matter of finding a good api.

```
const extendFilter = (logger, newFilter) => {
  logger.filter = (...args) => logger.filter(newFilter(...args));
};

const deriveWithFilter = curry('helix-log.deriveWithFiler', (logger, newFilter) => {
  const nu = shallowclone(logger);
  extendFilter(nu, newFilter);
  return nu;
});
```

These functions fulfill the following criteria:

* They are generic & work for any logger or interface
* They require implementing the shallowclone trait (which IMHO is a good change anyways; we should probably just implement any of the ferrum traits that apply)
* The interfaces stay almost as simple as they are now
* They are more flexible: Using something like _deepsetdefault is possible; a simple shallow merge is possible to (and fairly easy)
* Their semantics are very simple: _deepsetdefault is possible (a lodash method exists) but opt-in. (This avoids nasty surprises when developers expect a simple shallow merge or slightly different semantics).

The only real drawback is that this cannot be nicely chained. This can be remedied by using ferrum's `pipe()` method for a slightly odder, but still functional syntax.

Cases where setting fields or even deep merging fields is frequently required; in these cases the hassle of creating a custom filter function might be a bit much. In these cases a helper in application code can be used:

```
const deriveWithDefaultFields = (logger, fields) => deriveWithFilter((msg) => ({...msg, ...fields}));
const deriveWithMergedFields = (logger, fields) => deriveWithFilter((msg) => _.merge(msg, fields)));
```

koraa commented:
To clarify: Moving child to InterfaceBase makes .child() available to all all helix-log Interfaces, but not to all loggers…
Having it in all Interfaces *is* a definite improvement!

tripodsan commented:
I think it should be easy for the developer to achieve 2 things:
1. defining default fields for any logger. that's why I added the `fields` option to `InterfaceBase`
2. creating a derived logger w/o using an extra library and with very simple call. I don't care if its:
    `log.child({ ... })` or `derrive(log, { ...}`. but similarly simple

Since you have a better understanding how this could fit into the existing code, please take over the issue and PR. thanks!

tripodsan commented:
> I generally won't accept a PR that just introduces a generic concept for just one of the classes;

btw, if you do not accept a PR, please use the github feature to request chances. otherwise someone else might review and accept and merge ;-). the _request changes_ is a strong veto. not creating a review is not.

koraa commented:
Well, I am trying to have a design discussion…I don't really care who ends up implementing the feature, but we will have to come up with a design that we both feel is a good solution; this way we will end up with a better solution than either of us could have come up with on their own.

This is a process which has led the two of us to design better APIs in the past; e.g. with the generic fields support.

>  defining default fields for any logger. that's why I added the fields option to InterfaceBase
creating a derived logger w/o using an extra library and with very simple call. I don't care if its: log.child({ ... }) or derrive(log, { ...}. but similarly simple

Yes; that is what deriveWithFilter and extendFilter are supposed to provide, while requiring the developer to be explicit about how the fields merging is performed.

```
new SimpleInterface({ filter: (msg) => ({foo: 42, bar: 23, ...msg}); // Define default fields
new SimpleInterface({ filter: (msg) => ({...msg, foo: 42, bar: 23}); // Define override fields
new SimpleInterface({ filter: (msg) => _.merge({foo: 42, bar: 23}, msg); // Recursively merge default new SimpleInterface({ filter: (msg) => ({foo: 42, bar: 23, ...msg}); // Define default fields

deriveWithFilter((my_logger, msg) => ({foo: 42, bar: 23, ...msg})); // Derive with default fields
```

I realize this is not quite as convenient as an additional fields proper and requires a slightly steeper learning curve, but it does keep the complexity of the logging interfaces at a minimum and is more explicit. There is a trade off here.

Do you feel like this interface can satisfy your requirements? Can you provide examples of your use cases where this does not suffice or where the slightly increased complexity poses a serious problem?

tripodsan commented:
> I realize this is not quite as convenient as an additional fields proper and requires a slightly steeper learning curve, but it does keep the complexity of the logging interfaces at a minimum and is more explicit. There is a trade off here.

which I think is not appropriate :-) Since the _fields_ is a fundamental feature of the `InterfaceBase` (https://github.com/adobe/helix-log/blob/b828a7adb8f1f05cebd032e375e871af70a433c1/src/log.js#L765), I think it should also offer simple ways to manipulate them. eg by having the extra `fields` property.

> Do you feel like this interface can satisfy your requirements?

which one? mine, yes. the ambiguous filter, no. ;-)

> Can you provide examples of your use cases where this does not suffice or where the slightly increased complexity poses a serious problem?

I think even the current approach with the filter function would suffice.

koraa commented:
The only real contentious issue here seems to be whether a `defaultField` property should be introduced or not; we both agree that a `deriveWithDefaultFields` would be a beneficial addition.

I think at this point there is no real advantage to discussing the fields property any further, as you have stated your motivations (convenience) and I have stated mine (keeping our traits small/avoiding feature creep).

As you have stated already, the current implementation is sufficient for your use case so I stand by my opinion that we should at least defer adding the fields property until such a feature is needed in a lot of places, however I do not consider this to be sufficiently important to veto/block if you do decide to go ahead with the `fields` property…

koraa commented on src/log.js:
```diff
@@ -747,23 +747,55 @@ class MemLogger extends FormattedLoggerBase {
+   * todo: add to ferrum?
+   * @private
+   */
+  static _deepsetdefault(src, dst) {
```
Please use shallow merging only; deep merging/setdefault *should* be something that the developer actively opts into.

tripodsan replied:
hmm, it would be cool if we could support both, maybe with a dotted notation?

```
{
   'data.id': 42
}
```

koraa replied:
Sure we could! Although then we would have to think about the edge cases (e.g. what if a field containing a dot is supposed to be set? We would have to start coming up with an escaping scheme…possible but complex).

Or we could just have people use the filter field…in which case this feature is already supported…

During the initial fields support PR I wrote I started off with a fields property; I did want it to have features like deep merging, deleting fields, applying functions to fields and so on…I even have a file implementing a grammar very similar to the dot notation you are suggesting lying around; I at some point abandoned the approach because I realized we already have a DSL for transforming objects…javascript

I agree it would be cooler! I just do not think this coolness is worth the added complexity for helix-log and the increased learning curve for our users…

koraa commented on src/log.js:
```diff
@@ -747,23 +747,55 @@ class MemLogger extends FormattedLoggerBase {
   } = {}) {
-    assign(this, { logger, level, filter });
+    assign(this, {
+      logger, level, filter, fields,
```
Please use a more specific name; `defaultFields` instead of fields might be suitable…

Please introduce the fields property to all Loggers, not just the interfaces.

And finally please update the documentation of the Interface & Logger traits with the property;

(that they `SHOULD` (as in RFC should) support the property; if they do support the property they MUST perform a shallow merge/setdefault into the message AFTER applying the filters – This is to retain the ability for filters to set default parameters)

tripodsan replied:
I rather see this a functionality on the interface, not on the loggers.

koraa replied:
Loggers have the same basic interface as far as this feature is concerned; please, let's keep things consistent.

koraa commented on src/log.js:
```diff
@@ -747,23 +747,55 @@ class MemLogger extends FormattedLoggerBase {
+   * the given options.
+   * @param {object} opts Options to merge with this logger
+   */
+  child(opts) {
```
Let's please make this a function; we should also give it a more specific name (I still quite like deriveSomethingSomething); `child` is fairly nonspecific and invokes concepts from OOP)

tripodsan replied:
ok

koraa replied:
Still needs to be a function…

koraa commented on src/log.js:
```diff
@@ -792,6 +824,11 @@ class InterfaceBase {
  * @implements LoggingInterface
  */
 class SimpleInterface extends /* private */ InterfaceBase {
+  constructor({ fields = {}, ...opts } = {}) {
```
I do not think specifying fields here again will be necessary; the default in InterfaceBase should take care of that!

tripodsan replied:
good catch. probably a left over...

koraa commented:
I've left some minor feedback on the actual fields property implementation; mostly minor stuff, documentation. Generally the implementation looks quite fine!

tripodsan commented:
ok...here's an updated version....

koraa commented on src/log.js:
```diff
@@ -393,6 +393,10 @@ const __handleLoggingExceptions = (fields, logger, code) => {
+ * Loggers SHOULD provide a named constructor option 'defaultFields'; if they do support the
+ * property they MUST perform a shallow merge/setdefault into the message AFTER applying the
+ * filters.
+ *
```
This section also needs to be added to the LoggerInterface interface

tripodsan commented:
> Looks good! Feel free to merge without further review after both points from this review have been addressed!

I think, I addressed the issues. since I'm adding a new export, let's review once more (in case someone doesn't like the name)


https://github.com/adobe/helix-log/pull/67/files#diff-4f24bb7c46eda8f63690cb849ddb0f2dR344
